umbral/fragments.py

```diff
@@ -75,12 +75,79 @@ def __bytes__(self):
         return self.to_bytes()
 
 
+class CorrectnessProof(object):
+    def __init__(self, e2, v2, u1, u2, z1, z2, z3, metadata:bytes=None):
+        self.point_eph_e2 = e2
+        self.point_eph_v2 = v2
+        self.point_kfrag_commitment = u1
+        self.point_kfrag_pok = u2
+        self.bn_kfrag_sig1 = z1
+        self.bn_kfrag_sig2 = z2
+        self.bn_sig = z3
+        self.metadata = metadata
+
+    @classmethod
+    def from_bytes(cls, data: bytes, curve: ec.EllipticCurve=None):
+        """
+        Instantiate CorrectnessProof from serialized data.
+        """
+        curve = curve if curve is not None else default_curve()
+        key_size = get_curve_keysize_bytes(curve)
+        data = BytesIO(data)
+
+        # BigNums are the keysize in bytes, Points are compressed and the
+        # keysize + 1 bytes long.
+        e2 = Point.from_bytes(data.read(key_size + 1), curve)
+        v2 = Point.from_bytes(data.read(key_size + 1), curve)
+        kfrag_commitment = Point.from_bytes(data.read(key_size + 1), curve)
+        kfrag_pok = Point.from_bytes(data.read(key_size + 1), curve)
+        kfrag_sig1 = BigNum.from_bytes(data.read(key_size), curve)
+        kfrag_sig2 = BigNum.from_bytes(data.read(key_size), curve)
+        sig = BigNum.from_bytes(data.read(key_size), curve)
+
+        metadata = data.read()
+        if metadata == bytes(0):
+            metadata = None
+
+        return cls(e2, v2, kfrag_commitment, kfrag_pok, 
+                   kfrag_sig1, kfrag_sig2, sig, metadata=metadata)
+
+    def to_bytes(self) -> bytes:
+        """
+        Serialize the CorrectnessProof to a bytestring.
+        """
+        e2 = self.point_eph_e2.to_bytes()
+        v2 = self.point_eph_v2.to_bytes()
+        kfrag_commitment = self.point_kfrag_commitment.to_bytes()
+        kfrag_pok = self.point_kfrag_pok.to_bytes()
+        kfrag_sig1 = self.bn_kfrag_sig1.to_bytes()
+        kfrag_sig2 = self.bn_kfrag_sig2.to_bytes()
+        sig = self.bn_sig.to_bytes()
+
+        result = e2            \
+            + v2               \
+            + kfrag_commitment \
+            + kfrag_pok        \
+            + kfrag_sig1       \
+            + kfrag_sig2       \
+            + sig              
+
+        if self.metadata is not None:
+            result = result + self.metadata
```

Comment:
You can also do `result += self.metadata or b''`. This will concat the empty bytestring if `self.metadata` is `None` and removes the need for L135.

Reply:
Fixed